test/runtests.jl

```diff
@@ -1149,24 +1149,35 @@ end
                         ("'[4,4)'::int8range", Interval{Int64}()),  # Empty interval
-                        ("'[11.1,22.2)'::numrange", Interval(Decimal(11.1), Decimal(22.2), true, false)),
-                        ("'[2010-01-01 14:30, 2010-01-01 15:30)'::tsrange", Interval(DateTime(2010, 1, 1, 14, 30), DateTime(2010, 1, 1, 15, 30), true, false)),
-                        ("'[2010-01-01 14:30-00, 2010-01-01 15:30-00)'::tstzrange", Interval(ZonedDateTime(2010, 1, 1, 14, 30, tz"UTC"), ZonedDateTime(2010, 1, 1, 15, 30, tz"UTC"), true, false)),
-                        ("'[2004-10-19 10:23:54-02, 2004-10-19 11:23:54-02)'::tstzrange", Interval(ZonedDateTime(2004, 10, 19, 12, 23, 54, tz"UTC"), ZonedDateTime(2004, 10, 19, 13, 23, 54, tz"UTC"), true, false)),
-                        ("'[2004-10-19 10:23:54-02, Infinity)'::tstzrange", Interval(ZonedDateTime(2004, 10, 19, 12, 23, 54, tz"UTC"), ZonedDateTime(typemax(DateTime), tz"UTC"), true, false)),
-                        ("'(-Infinity, Infinity)'::tstzrange", Interval(ZonedDateTime(typemin(DateTime), tz"UTC"), ZonedDateTime(typemax(DateTime), tz"UTC"), false, false)),
-                        ("'[2018/01/01, 2018/02/02)'::daterange", Interval(Date(2018, 1, 1), Date(2018, 2, 2), true, false)),
+                        ("'[11.1,22.2)'::numrange", Interval{Closed, Open}(Decimal(11.1), Decimal(22.2))),
+                        ("'[2010-01-01 14:30, 2010-01-01 15:30)'::tsrange", Interval{Closed, Open}(DateTime(2010, 1, 1, 14, 30), DateTime(2010, 1, 1, 15, 30))),
+                        ("'[2010-01-01 14:30-00, 2010-01-01 15:30-00)'::tstzrange", Interval{Closed, Open}(ZonedDateTime(2010, 1, 1, 14, 30, tz"UTC"), ZonedDateTime(2010, 1, 1, 15, 30, tz"UTC"))),
+                        ("'[2004-10-19 10:23:54-02, 2004-10-19 11:23:54-02)'::tstzrange", Interval{Closed, Open}(ZonedDateTime(2004, 10, 19, 12, 23, 54, tz"UTC"), ZonedDateTime(2004, 10, 19, 13, 23, 54, tz"UTC"))),
+                        ("'[2004-10-19 10:23:54-02, Infinity)'::tstzrange", Interval{Closed, Open}(ZonedDateTime(2004, 10, 19, 12, 23, 54, tz"UTC"), ZonedDateTime(typemax(DateTime), tz"UTC"))),
+                        ("'(-Infinity, Infinity)'::tstzrange", Interval{Open, Open}(ZonedDateTime(typemin(DateTime), tz"UTC"), ZonedDateTime(typemax(DateTime), tz"UTC"))),
+                        ("'[2018/01/01, 2018/02/02)'::daterange", Interval{Closed, Open}(Date(2018, 1, 1), Date(2018, 2, 2))),
+                        # Unbounded ranges
+                        ("'[3,)'::int4range", Interval(Int32(3), nothing)),
+                        ("'[3,]'::int4range", Interval(Int32(3), nothing)),
+                        ("'(3,)'::int4range", Interval{Closed, Unbounded}(Int32(4), nothing)),
+                        ("'(,3)'::int4range", Interval{Unbounded, Open}(nothing, Int32(3))),
+                        ("'(,3]'::int4range", Interval{Unbounded, Open}(nothing, Int32(4))),
```

Comment:
Another invalid test:
```suggestion
                        ("'(,3]'::int4range", Interval{Unbounded, Closed}(nothing, 3)),
```

Reply:
This is how postgres stores that in the db. Changing it to your suggestion doesn't work:

```
(test_str, data) = ("'(,3]'::int4range", Interval{Int32,Unbounded,Closed}(nothing, 3)): Test Failed at /Users/nicoleepp/.julia/dev/LibPQ/test/runtests.jl:1191
  Expression: isequal(parsed, data)
   Evaluated: isequal(Interval{Int32,Unbounded,Open}(nothing, 4), Interval{Int32,Unbounded,Closed}(nothing, 3))
```

Reply:
Very interesting. This seems to only apply to the upper endpoint.
```
omus=# SELECT '(,3]'::int4range;
 int4range
-----------
 (,4)
(1 row)
```
I would suggest at least adding a comment about this as it is rather strange. I'll see if I can find out more about this.
```suggestion
                        ("'(,3]'::int4range", Interval{Unbounded, Open}(nothing, Int32(4))),  # Postgres automatically updates `(,3]` to `(,4)`
```

Reply:
I didn't find anything official but I believe what is going on here is that PostgreSQL is normalizing the range (see the last example in https://www.postgresql.org/docs/9.3/rangetypes.html#RANGETYPES-IO).
```
omus=# SELECT '[4,4)'::numrange;
 numrange
----------
 empty
```
Since we're using `int4range` here PostgreSQL can change the interval as it knows the possible values that are included in the interval. We can see this in action by making an interval set that just contains `4`:
```
omus=# SELECT '[4,4]'::int4range;
 int4range
-----------
 [4,5)
(1 row)

omus=# SELECT '(3,4]'::int4range;
 int4range
-----------
 [4,5)
(1 row)

omus=# SELECT '[4,5)'::int4range;
 int4range
-----------
 [4,5)
(1 row)

omus=# SELECT '(3,5)'::int4range;
 int4range
-----------
 [4,5)
(1 row)
```
Finally, for ranges which the sets are not so clearly defined no normalization should occur which is the case:
```
omus=# SELECT '[4.0,4.0]'::numrange;
 numrange
-----------
 [4.0,4.0]
(1 row)
```